### ISSUE TITLE
Allow composition of notation

### ISSUE THREAD

**Issue by keenancrane** [penrose/penrose#764]
Even in our original sugared notation system, there was limited (or no?) support for composition of notation.

As we re-implement notation (see #763), it would be great to support composition of notation—just like we support composition of unsugared Substance.

A good example use case is matrix operations.  Consider for instance some Substance statements
```
Matrix A
Matrix B := Inverse(Transpose(A))
```
It's natural to want to notate this second expression as something like
```
Matrix B = (Aᵀ)⁻¹,
```
where the Unicode `⁻¹` denotes the inverse, and the Unicode `ᵀ` denotes the transpose.

One issue to resolve is how to specify (and resolve) operator precedence when operations aren't expressed/grouped using explicit parentheses.  From discussion on Slack with @keenancrane and @JonathanAldrich:

---

@keenancrane: One small item to add to that agenda is order of operations, which currently is undefined (right?) in our sugaring semantics. For instance if I define
```
notation “x*y” ~ “multiply(x,y)”
notation “x+y” ~ “add(x,y)”
```
It’s not clear what the Substance expression `u+v*w` should evaluate to: is it `add(u,multiply(v,w))` or is it `multiply(add(u,v),w)`? Of course this was never an issue before since sugared notation wasn’t composable. But feels like it really should be part of the language spec. In particular, there should be a way to specify a partial ordering on the precedence of operators (I say partial because not all operators have compatible types.) One possibility is to infer the order from the line number, but this feels brittle—suppose you copy and paste code from another Domain schema, or in some future version of Penrose we allow cascading schemas. An alternative is to explicitly specify an ordinal precedence as a number, e.g.,
```
notation “x*y” ~ “multiply(x,y)” 1
notation “x+y” ~ “add(x,y)” 2
```
indicating that multiplication binds tighter than addition. Alternatively, one could specify a relative ordering (like our “layer” keyword in Style).

Related to this question: how do we [want to] treat parentheses in Substance? Does every domain include parentheses by default? Parentheses are critical to resolving operator precedence, via `(u+v)*w vs. u+(v*w)` (even when a default precedence is well-specified). But not all mathematical notation uses parentheses in this way. For instance, a tuple is often `(x,y,z)`, and some binary operators like the inner product are sometimes notated via `(u,v)`. Some people also use a melange of brackets for nesting within the same expression, e.g., { [ and ( for different levels of nesting. But ok, we don’t have to permit everything under the sun!

Does this issue pop up in other languages (like Coq, Lean, etc.)?

---

@JonathanAldrich: I don't know how the Coq mechanism works.  I do know that it's a hack at least to some extent, because it's not possible to compose arbitrary grammars without the possibility to get conflicts.

For Penrose, we have a few choices:
- Build in some key notations that we think are important (e.g. addition, multiplication)
- Support extension, but only in limited enough ways that we can guarantee there aren't conflicts
- Support arbitrary extension, and try to have the inevitable failures detected with nice warning messages...recognizing that those warning messages may still come at unexpected times (e.g. loading two Penrose libraries that work fine separately but not together)
- Give up on usability for extensions

Obviously I prefer not to choose the last of these. 

---

@keenancrane: Wow, I’m surprised it’s this difficult. Is it not standard to specify operator precedence in a language?

---

@JonathanAldrich: If all you are doing is operator precedence, and you can agree up front on what that is, it is easy.  What's hard is adding random other notation and expecting it to work.  Or, having different operator extensions that want different precedences.

If you define the # binary operator to have precedence above +, and Josh defines the # binary operator to have precedence below +, then if I write a substance program that wants to use both definitions, it won't work.

---

@keenancrane: I see. But provided there is only a single Domain program, with a unique ordinal assigned to each operator, sounds like it should be ok to resolve. (And the compiler can barf otherwise.)

Is there a simple example of “other notation”? I guess I’m just thinking about unary, binary, maybe ternary operators (as defined by Domain functions).

I guess the type signatures also help to resolve some ambiguities. (Though may be costly to do the analysis for long expressions?)

---

@JonathanAldrich: True that if there is only one Domain program this is OK (and actually the very worst problems all get easier in this case).  I suspect that assumption will be limiting sooner rather than later, but we are not there yet.

Also yes, types may help to disambiguate...using them would probably require thinking about theory a little bit, but this is a relatively small thing (maybe couple of weeks, not 3 months).

Types scale pretty well so the length of expression is not so important.

Examples: from the codebase , see geometry - we proposed notation `"∠(p, q, r)" ~ "InteriorAngle(p, q, r)"`   Not technically an operator, although anything with parentheses is not too hard if you build in parenthesis matching somehow.

This one from linear algebra is weird because it's not even a full statement: `notation "Scalar c := " ~ "Scalar c; c := "`

But most of the ones we've actually prototyped in the old version (where sugar worked) are not that hard.  If our ambitions don't go too much beyond that we can probably get a practical system work, we just need to think a bit about what the boundaries are, otherwise we risk people trying to push it and starting to get weird and unpleasant behavior :slightly_smiling_face:

---

@keenancrane: Yeah, exactly.  Prioritizing composability over absolute generality may put some constraints on what notation can look like—but those constraints may be helpful to compilers and programmers alike.

**Comment by christianp** [comment]
If you're looking for examples of notation that is hard to parse, I keep a collection of ambiguities and contradictory conventions at [whystartat.xyz](https://whystartat.xyz/wiki/Main_Page).

An example of notation that uses brackets/parentheses completely differently is interval notation. There are two common conventions: in the first, you use parentheses for an open end and square brackets for a closed end; in the other, you use inward-facing square brackets for a closed end (e.g. `[a,b]`) and outward-facing square brackets for an open end (e.g. `]a,b[`). I don't think it would be feasible to allow this!

In case it helps, I have a notational wishlist for the parser in Numbas, at numbas/Numbas#677. My plan at the moment, since there are so many mutually-exclusive notations, is to have some syntax to explicitly switch to a given notation.

**Comment by samestep** [comment]
Thanks @christianp! This looks useful for deciding how to scope the generality of our notation system.

**Comment by samestep** [comment]
Closing in favor of https://github.com/penrose/penrose/issues/763#issuecomment-1422869957.
